File: day1/cuwano/day1.hs
sum' :: (Num a, Eq a) => [a] -> a
sum' xs@(x:_) = megasum xs x

megasum :: (Num a, Eq a) => [a] -> a -> a
megasum [x] y
  | x == y = x
  | otherwise = 0
megasum (x:y:xs) z
  | x == y = x + megasum (y:xs) z
  | otherwise = megasum (y:xs) z

File: day1/cuwano/day1.ex
defmodule AdventCode do
  def sum([x|_] = xs) do
    megasum xs, x
  end

  def megasum([x,x|xs], h), do: x + megasum([x|xs], h)
  def megasum([_,x|xs], h), do: megasum([x|xs], h)
  def megasum([x], x), do: x
  def megasum(_, _), do: 0
end
